src/ZODB/Connection.py

```diff
@@ -1300,6 +1300,23 @@ def __repr__(self):
 size.
 """
 
+def _extension_property(func):
+    import inspect
```

Comment:
Imports at top-of-module for non-test code please.  Also, this descriptor factory needs a docstring explaining just what it does.

Reply:
> Imports at top-of-module for non-test code please.

This function is only called at load-time so I imported locally to not pollute globals. But if you prefer, I'll change.

> this descriptor factory needs a docstring explaining just what it does.

Ouch, not easy. What about:
```py
def _extension_property(func):
    """
    When called twice, this builds a pair of properties that maintain a value
    in 2 different forms, in such a way that conversion between the 2 is done
    lazily. The 2 decorated functions must be inverse of each other:
      def foo(bar): ...
      def bar(foo): ...
      where foo(bar(x)) == x

    The name of their parameter specifies that of the mirror attribute.

    This helps to make sure that 2 properties remain always symmetric.
    Currently used to automatically convert transaction extension from/to bytes.
    """
```

Reply:
The main goal of this PR is to avoid (de)serialize whenever possible, because `dumps(loads(x))` is not always equal to `x` and we want to be able to compare bit-to-bit (longer explanation in the PR description).

Reply:
If `dumps` and `loads` are not complements, then the whole use case blows up.  Store the data on disk in one format, sure, but keep it in only one *real* attribute while in RAM, and let the other one be computed logically from it (I don't really care whether `extension` or `extension_bytes` is the canonical one.

Reply:
> If `dumps` and `loads` are not complements, then the whole use case blows up.

I don't understand.

> I don't really care whether `extension` or `extension_bytes` is the canonical one.

If `extension_bytes` is the canonical one, yes it works. But from a performance point of view, that looks horrible, since you'd convert whenever you access/modify `extension` (isn't it the main way extensions are manipulated ?)

Reply:
Couldn't this be broken anyway? `extension` is a `dict`, and I would guess if there's any mutation of it, it's in-place, not by setting a whole new `dict` on the property.

Reply:
Arg, I missed that. The only way to solve that while still storing extension_bytes verbatim, is to delete the cached extension_bytes whenever extension is accessed. I must change `_extension_property`.

Reply:
I mean:
```diff
@@ -1309,7 +1322,7 @@ def fget(self):
             return self.__dict__[name]
         except KeyError:
             d = self.__dict__
-            value = d[name] = func(d[alt])
+            value = d[name] = func(d.pop(alt))
             return value
     def fset(self, value):
         d = self.__dict__
```

Reply:
Sorry, now that I'm thinking about it, that still seems a bit foot-gunny to me. If the intent is to be able to copy a database using the iterator APIs byte-for-byte identical, then if *anything* in that process touches `.extension` we'd have to reserialize, and that would defeat the purpose. 

Maybe a big-ole disclaimer in the docs is enough to mitigate that?

Reply:
But code doing bit-to-bit checks is unlikely to access `.extension`, isn't it ?

With my suggested fixup, the PR preserves the original behaviour, in that any change to .extension is "saved" (i.e. serialized), but that also does not look important: I'd be surprised to see code relying on that.

My feeling is that either behaviour is fine, but whatever we choose (between too much or not enough reserialization), I suppose we can't avoid documenting.